### PR TITLE
Climb structure

### PR DESCRIPTION
**Pull request by NoamRosenberg08**

### CODE REVIEW

File: src/main/java/frc/robot/superstructure/climb/ClimbConstants.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package frc.robot.superstructure.climb;
+
+public class ClimbConstants {
+
+	public static final double SOLENOID_RELEASE_TIME = 0.3;
+	public static final double SOLENOID_RETRACTING_UNTIL_HOLDING_TIME = 0.2;
+
+
```
Comment (YHerm) [inline src/main/java/frc/robot/superstructure/climb/ClimbConstants.java:8]: remove space

File: src/main/java/frc/robot/superstructure/climb/ClimbState.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package frc.robot.superstructure.climb;
+
+public enum ClimbState {
+
+	RETRACT,
+	EXTEND,
+	STOP
```
Comment (YHerm) [inline src/main/java/frc/robot/superstructure/climb/ClimbState.java:7]: add space

File: src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	public Command setState(ClimbState state) {
+		return switch (state) {
+			case STOP -> lifterStateHandler.setState(LifterState.STOP).alongWith(solenoidStateHandler.setState(SolenoidState.OFF));
+			case EXTEND ->
+					(lifterStateHandler.setState(LifterState.BACKWARD)
+					.raceWith(new WaitCommand(ClimbConstants.SOLENOID_RELEASE_TIME)))
+					.andThen((solenoidStateHandler.setState(SolenoidState.RETRACT))
+					.raceWith(new WaitCommand(ClimbConstants.SOLENOID_RETRACTING_UNTIL_HOLDING_TIME)))
+					.andThen(lifterStateHandler.setState(LifterState.EXTENDED).alongWith(solenoidStateHandler.setState(SolenoidState.HOLD)));
+			case RETRACT -> lifterStateHandler.setState(LifterState.RETRACTED).alongWith(solenoidStateHandler.setState(SolenoidState.OFF));
```
Comment (YHerm) [inline src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java]: split to funcs

**Review by YHerm — COMMENTED** [review]
merge core

File: src/main/java/frc/robot/JoysticksBindings.java [hunk 4]
```diff
@@ -23,6 +27,15 @@ public static void configureBindings(Robot robot) {
 
 	private static void mainJoystickButtons(Robot robot) {
 		SmartJoystick usedJoystick = MAIN_JOYSTICK;
+
+		ClimbStateHandler climbStateHandler = new ClimbStateHandler(
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/JoysticksBindings.java]: Shouuld this be here?

File: src/main/java/frc/robot/JoysticksBindings.java [hunk 5]
```diff
@@ -23,6 +27,15 @@ public static void configureBindings(Robot robot) {
...
+			new LifterStateHandler(robot.getLifter()),
+			new SolenoidStateHandler(robot.getSolenoid())
+		);
+
+		usedJoystick.B.onTrue(climbStateHandler.setState(ClimbState.STOP));
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/JoysticksBindings.java]: Delete test buttons

File: src/main/java/frc/robot/subsystems/lifter/Lifter.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	public LifterCommandsBuilder getCommandsBuilder() {
+		return lifterCommandsBuilder;
+	}
+
+	@Override
+	protected void subsystemPeriodic() {
+		updateInputs();
+		Logger.recordOutput("lifter position", convertToMeters(lifterStuff.positionSignal().getLatestValue()));
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/lifter/Lifter.java]: "Lifter position in meters"

File: src/main/java/frc/robot/subsystems/lifter/LifterCommandsBuilder.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import edu.wpi.first.wpilibj2.command.FunctionalCommand;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+
+import java.util.function.DoubleSupplier;
+
+public class LifterCommandsBuilder {
+
+	private final Lifter lifter;
+
+
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/lifter/LifterCommandsBuilder.java]: Delete extra apace line

File: src/main/java/frc/robot/subsystems/lifter/LifterCommandsBuilder.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	public LifterCommandsBuilder(Lifter lifter) {
+		this.lifter = lifter;
+	}
+
+	public Command setPower(double power) {
+		return new FunctionalCommand(
+				() -> {},
+				() -> lifter.setPower(power),
+				(interrupted) -> lifter.stop(),
+				() -> false, lifter
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/lifter/LifterCommandsBuilder.java]: Go down a row

File: src/main/java/frc/robot/subsystems/lifter/LifterConstants.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package frc.robot.subsystems.lifter;
+
+public class LifterConstants {
+
+	public static final String LOG_PATH = "lifter/";
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/lifter/LifterConstants.java]: Capital L, and Subsystems/

File: src/main/java/frc/robot/subsystems/lifter/LifterConstants.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+	public static final double EXTENDING_POWER = 0.9;
+	public static final double RETRACTING_POWER = -0.9;
+
+	public static final double FORWARD_POWER = 0.2;
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/lifter/LifterConstants.java]: What are those?

File: src/main/java/frc/robot/subsystems/lifter/LifterConstants.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package frc.robot.subsystems.lifter;
+
+public class LifterConstants {
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/lifter/LifterConstants.java]: I prefer space lines and souble space lines when needed

File: src/main/java/frc/robot/subsystems/lifter/LifterState.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+	HOLD(0, 0),
+	FORWARD(0, 0.2),
+	BACKWARD(0, -0.2),
+	RETRACTED(0.05, -0.9),
+	EXTENDED(0.25, 0.9);
+
+	private final double targetPosition;
+	private final double power;
+
+
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/lifter/LifterState.java]: Delete

File: src/main/java/frc/robot/subsystems/lifter/factory/LifterRealConstants.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+import frc.robot.subsystems.lifter.LifterStuff;
+import frc.utils.AngleUnit;
+
+import static edu.wpi.first.math.util.Units.inchesToMeters;
+import static edu.wpi.first.units.Units.Seconds;
+import static edu.wpi.first.units.Units.Volts;
+
+public class LifterRealConstants {
+
+	private static final double DRUM_RADIUS = inchesToMeters(0.96);
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/lifter/factory/LifterRealConstants.java]: Space

Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/lifter/factory/LifterRealConstants.java]: _METERS

File: src/main/java/frc/robot/subsystems/lifter/factory/LifterRealConstants.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+	public static final int MOTOR_CONFIGURATION_TRIES = 5;
+
+	private static TalonFXConfiguration generateMotorConfiguration() {
+		TalonFXConfiguration configuration = new TalonFXConfiguration();
+
+		configuration.Feedback.SensorToMechanismRatio = 7 * (60.0 / 24.0);
+
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/lifter/factory/LifterRealConstants.java]: Delete

File: src/main/java/frc/robot/subsystems/lifter/factory/LifterRealConstants.java [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+		return new LifterStuff(
+			logPath,
+			new TalonFXMotor(logPath, talonFXWrapper, generateSysidConfig()),
+			DRUM_RADIUS,
+			Phoenix6SignalBuilder
+				.generatePhoenix6Signal(talonFXWrapper.getPosition(), GlobalConstants.DEFAULT_SIGNALS_FREQUENCY_HERTZ, AngleUnit.ROTATIONS)
+		);
+	}
+
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/lifter/factory/LifterRealConstants.java]: Delete

File: src/main/java/frc/robot/subsystems/solenoid/SolenoidState.java [hunk 16]
```diff
@@ -0,0 +1,20 @@
+package frc.robot.subsystems.solenoid;
+
+public enum SolenoidState {
+
+	OFF(0),
+	RETRACT(0.8),
+	HOLD(0.2);
+
+
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/solenoid/SolenoidState.java:10]: Delete

File: src/main/java/frc/robot/subsystems/solenoid/SolenoidStateHandler.java [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+	private final Solenoid solenoid;
+
+	public SolenoidStateHandler(Solenoid solenoid) {
+		this.solenoid = solenoid;
+	}
+
+	public Command setState(SolenoidState state) {
+		return solenoid.getCommandsBuilder().setPower(state.getAppliedPower());
+	}
+
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/solenoid/SolenoidStateHandler.java]: Delete

File: src/main/java/frc/robot/superstructure/climb/ClimbConstants.java (hunk 1 above)
Comment (bossyshrimp68) [inline src/main/java/frc/robot/superstructure/climb/ClimbConstants.java:8]: Delete

File: src/main/java/frc/robot/superstructure/climb/ClimbConstants.java [hunk 18]
```diff
@@ -0,0 +1,9 @@
+package frc.robot.superstructure.climb;
+
+public class ClimbConstants {
+
+	public static final double SOLENOID_RELEASE_TIME = 0.3;
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/superstructure/climb/ClimbConstants.java]: Space

Comment (bossyshrimp68) [inline src/main/java/frc/robot/superstructure/climb/ClimbConstants.java]: _SECONDS

File: src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+            case EXTEND -> (
+                    lifterStateHandler.setState(LifterState.BACKWARD)
+                            .raceWith(new WaitCommand(ClimbConstants.SOLENOID_RELEASE_TIME))
+            ).andThen(
+                            (solenoidStateHandler.setState(SolenoidState.RETRACT))
+                                    .raceWith(new WaitCommand(ClimbConstants.SOLENOID_RETRACTING_UNTIL_HOLDING_TIME))
+                    )
+                    .andThen(
+                            lifterStateHandler.setState(LifterState.EXTENDED)
+                                    .alongWith(solenoidStateHandler.setState(SolenoidState.HOLD).raceWith(new WaitCommand(7)).andThen(solenoidStateHandler.setState(SolenoidState.OFF)))
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java]: 7?

File: src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+                    );
+            case RETRACT ->
+                    lifterStateHandler.setState(LifterState.RETRACTED).alongWith(solenoidStateHandler.setState(SolenoidState.OFF));
+        };
+    }
+
+
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java:32]: Delete

File: src/main/java/frc/robot/subsystems/solenoid/SolenoidState.java [hunk 21]
```diff
@@ -0,0 +1,20 @@
+package frc.robot.subsystems.solenoid;
+
+public enum SolenoidState {
+
+	OFF(0),
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/subsystems/solenoid/SolenoidState.java]: Wha5 does off mean

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]
"Delete" meas delete the extra space line. Did this go through format?

File: src/main/java/frc/robot/subsystems/lifter/factory/LifterRealConstants.java (hunk 13 above)
Comment (NoamRosenberg08) [reply src/main/java/frc/robot/subsystems/lifter/factory/LifterRealConstants.java]: where?

File: src/main/java/frc/robot/superstructure/climb/ClimbConstants.java (hunk 18 above)
Comment (NoamRosenberg08) [reply src/main/java/frc/robot/superstructure/climb/ClimbConstants.java]: where?

File: src/main/java/frc/robot/subsystems/solenoid/SolenoidState.java (hunk 21 above)
Comment (NoamRosenberg08) [reply src/main/java/frc/robot/subsystems/solenoid/SolenoidState.java]: not axtivating anything, like not holding and not retracting. when the solenoid is using 0 power

File: src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+	}
+
+	public Command setState(ClimbState state) {
+		return switch (state) {
+			case STOP -> lifterStateHandler.setState(LifterState.HOLD).alongWith(solenoidStateHandler.setState(SolenoidState.OFF));
+			case EXTEND ->
+				(lifterStateHandler.setState(LifterState.BACKWARD).raceWith(new WaitCommand(ClimbConstants.SOLENOID_RELEASE_TIME_SECONDS)))
+					.andThen(
+						(solenoidStateHandler.setState(SolenoidState.RETRACT))
+							.raceWith(new WaitCommand(ClimbConstants.SOLENOID_RETRACTING_UNTIL_HOLDING_TIME))
```
Comment (YHerm) [inline src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java]: u can do with timeout

File: src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+					)
+					.andThen(
+						lifterStateHandler.setState(LifterState.EXTENDED)
+							.alongWith(
+								solenoidStateHandler.setState(SolenoidState.HOLD)
+									.raceWith(new WaitCommand(ClimbConstants.SOLENOID_TIMEOUT_SECONDS))
```
Comment (YHerm) [inline src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java]: with timeout

File: src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+		this.lifterStateHandler = lifterStateHandler;
+		this.solenoidStateHandler = solenoidStateHandler;
+	}
+
+	public Command setState(ClimbState state) {
+		return switch (state) {
+			case STOP -> lifterStateHandler.setState(LifterState.HOLD).alongWith(solenoidStateHandler.setState(SolenoidState.OFF));
+			case EXTEND ->
+				(lifterStateHandler.setState(LifterState.BACKWARD).raceWith(new WaitCommand(ClimbConstants.SOLENOID_RELEASE_TIME_SECONDS)))
+					.andThen(
```
Comment (YHerm) [inline src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java:24]: maybe there are more cleanings to this command more than timeouts (i dont have an idea now)

Comment (NoamRosenberg08) [reply src/main/java/frc/robot/superstructure/climb/ClimbStateHandler.java:24]: well, there is, ill do it without timouts

**Review by YHerm — DISMISSED** [review]
merge all others before

File: src/main/java/frc/robot/subsystems/lifter/LifterStateHandler.java [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+
+	public LifterStateHandler(Lifter lifter) {
+		this.lifter = lifter;
+	}
+
+	public Command setState(LifterState state) {
+		return switch (state) {
+			case HOLD -> lifter.getCommandsBuilder().stop();
+			case FORWARD -> lifter.getCommandsBuilder().setPower(state.getPower());
+			case BACKWARD -> lifter.getCommandsBuilder().setPower(state.getPower());
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/subsystems/lifter/LifterStateHandler.java]: cant you do
```suggestion
			case FORWARD, BACKWARD -> lifter.getCommandsBuilder().setPower(state.getPower());
```
?

**Pull request merged**